### PR TITLE
[PT-D][3/N] Add FSDP hook with Named Optimizer

### PR DESCRIPTION
**Pull request by fduwjj**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #91321

### CODE REVIEW

**Comment by fduwjj** [comment]
Still WIP, please don't review it. Thanks!

File: test/distributed/fsdp/test_fsdp_optim_state.py [hunk 1]
```diff
@@ -1475,6 +1476,9 @@ def forward(self, x):
             loss = model(batch).sum()
             loss.backward()
             optim.step()
+            # if isinstance(optim, _NamedOptimizer):
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_optim_state.py:1479]: Not sure what this is for?

File: torch/distributed/optim/named_optimizer.py [hunk 2]
```diff
@@ -60,6 +62,7 @@ def __init__(
         named_parameters: Mapping[str, Union[torch.Tensor, ShardedTensor]],
         optimizer_class: optim.Optimizer,
         param_groups: Collection[Mapping[str, Any]] = None,
+        module: nn.Module = None,
```
Comment (rohan-varma) [inline torch/distributed/optim/named_optimizer.py:65]: docstring

File: torch/distributed/optim/named_optimizer.py [hunk 3]
```diff
@@ -268,6 +273,16 @@ def add_param_group(self, param_group: Any) -> None:
             "add_param_group not supported yet and might be implemented soon."
         )
 
+    def _pre_load_state_dict(self, state_dict) -> None:
+        if isinstance(self.module, FSDP):
+            return FSDP._load_optim_state_dict_pre_hook(self.module, self._optimizer, state_dict)
```
Comment (rohan-varma) [inline torch/distributed/optim/named_optimizer.py:278]: why are we returning if it is FSDP, shouldn't we do the preprocessing?

Comment (fduwjj) [reply torch/distributed/optim/named_optimizer.py:278]: `_load_optim_state_dict_pre_hook` does the preprocessing here.

File: test/distributed/fsdp/test_fsdp_optim_state.py (hunk 1 above)
Comment (fduwjj) [reply test/distributed/fsdp/test_fsdp_optim_state.py:1479]: opps. Let me update the PR. These were removed already.

**Review by fegin — APPROVED** [review]
LGTM

**Comment by fduwjj** [comment]
Rebased multiple times and failures are different every time, so I think it's not related to this PR. Feel free to revert it, if this PR is the culprit.

**Pull request closed without merging**